### DISCUSSION TITLE
Nested Comments schema and query

### DISCUSSION THREAD

**Discussion by viperfx** [prisma/prisma#4394] [Q&A]
I have a use case where I need to do nested comments. **Only 2 levels deep.**. I have the following schema that I use, where comments can have a parentComment and can be used to nest the comments for rendering.

```prisma
model Comment {
  id              String              @id @default(cuid())
  user            User                @relation(fields: [userId], references: [id])
  userId          Int
  comment         String?
  parentCommentId String?
  parentComment   RoadmapItemComment? @relation("CommentParent", fields: [parentCommentId], references: [id])
}

model Post {
  name        String
  slug        String
  desc        String
  comments    RoadmapItemComment[]
  ....
}
```

The following is used to fetch the data

```ts
export const roadmapItem = ({...}) => {
  requireAuth()
  return db.Post.findOne({
    where: {
      id: ...
    },
    include: {
      comments: {
        include: {
          user: true,
        }
      }
    }
  })
}
```

This gets a flat structure of comments, and I use some lodash functions to nest the comments. With the following function:

```ts
const deflattenComments = (comments) => {
  const childComments = filter(comments, (o) => o.parentCommentId != null)
  const parentComments = filter(comments, {parentCommentId: null});
  function findReply(c) {
    return {
      ...c,
      comments: filter(childComments, {parentCommentId: c.id})
    };
  }
  return map(parentComments, findReply);
}
```

This works well for now, but I see some issues:

- It may not scale well as the number of comments grows larger.
- Doing infinite loading or pagination may not work, due to child and parent comments parents fetched in order of creation. 

Is there a better way to model and query a 2 level Comment system?

**Comment by sehmaschine** [comment]
I'm not sure if this works, but couldn't you extend the `Comment` model with an attribute `childComments` (relation to `parentCommentId`), then filter on parent comments first (within your where clause) and `childComments` as an include?
